Ctrl-C management in here_doc plus print echo $?

dbejar-s opened the pull request.

dbejar-s committed:
Ctrl-C management in here_doc plus print echo $?

pmarkaide commented on src/general_utils.c:
```diff
@@ -94,5 +94,8 @@ t_macro	*init_macro(char **envp, char **argv)
 	macro->pipe_fd[0] = -1;
 	macro->pipe_fd[1] = -1;
 	macro->read_end = -1;
+	macro->exit_code = 0;
```
Not needed to set to zero if we calloc, supposatelly

pmarkaide commented on src/main.c:
```diff
@@ -20,7 +20,18 @@ static char	*read_line(t_macro *macro)
+		macro->here_doc_flag = 0;
+	}
 	line = readline(path);
+	signal(SIGINT, sigint_handler_in_parent);
```
Nice

pmarkaide commented on src/signals.c:
```diff
@@ -39,3 +40,21 @@ void	sigquit_handler_in_child(int sig)
 	}
 }
+
+void	sigint_handler_after_here_doc(int sig)
```
Isn't this just the same as siging_handler_in_parent?

dbejar-s replied:
Difference is that I don't print a new line... I just copied my bash behavior. Let's check it out.

pmarkaide commented on src/parsing_utils.c:
```diff
@@ -48,7 +51,21 @@ static int	read_here_doc(t_token *token, t_macro *macro)
 		line = readline("> ");
+		if (g_exit == SIGINT)
+		{
+			open("/dev/tty", O_RDONLY);
```
can you explain what /dev/tty does?

pmarkaide commented on src/signals.c:
```diff
@@ -39,3 +40,21 @@ void	sigquit_handler_in_child(int sig)
+{
+	if (sig == SIGINT)
+	g_exit = SIGINT;
+	close(STDIN_FILENO);
```
This is clever, I couldn't stop the deadline sending EOF or anything

dbejar-s replied:
I tried so many different ways to make readline stops... and I knew it needs EOF and the only way I figured out to "fake" a EOF is by closing the STDIN. Problem is that you need to open the STDIN as soon as you're done with the exit procedure caused by SIGINT, hence the open("/dev/tty", O_RDONLY)... which is basically a way to undo the close(STDIN_FILENO). Thing is that opening /dev/tty ignores any other fd redirection that is in place... but this only happens with here_doc, so it should be ok

pmarkaide replied:
I tried using `ioctl(STDIN_FILENO, TIOCSTI, "\004");` that sends EOF to the terminal, but it did not close deadline. I will ask in school

pmarkaide reviewed (commented):
I've tested and it works. Is clear and I like the implementation design. We need to add a newline also in the here_doc_handler to make it as bash, but apart from that seem to work. Let's discuss details in the meeting and we merge

dbejar-s committed:
fix: remove unnecessary free_string() call in read_here_doc()

The commit removes an unnecessary call to the free_string() function in the read_here_doc() function in parsing_utils.c. This call was freeing the 'del' string twice, which is not needed and can cause memory issues. The removal of this call improves the code's efficiency and avoids potential memory leaks.
